tls/s2n_async_pkey.c

```diff
@@ -409,3 +422,140 @@ S2N_RESULT s2n_async_pkey_sign_free(struct s2n_async_pkey_op *op)
+    struct s2n_blob *in = &decrypt->encrypted;
+
+    *data_len = in->size;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_size_sign(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    struct s2n_hash_state *digest = &sign->digest;
+
+    uint8_t digest_length = 0;
+    RESULT_GUARD_POSIX(s2n_hash_digest_size(digest->alg, &digest_length));
+
+    *data_len = digest_length;
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_async_pkey_op_get_input(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(data);
+
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    POSIX_GUARD_RESULT(s2n_async_get_actions(op->type, &actions));
+    POSIX_ENSURE_REF(actions);
+
+    POSIX_GUARD_RESULT(actions->get_input(op, data, data_len));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_decrypt(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_decrypt_data *decrypt = &op->op.decrypt;
+    struct s2n_blob *in = &decrypt->encrypted;
+
+    RESULT_ENSURE_LTE(in->size, data_len);
+
+    RESULT_CHECKED_MEMCPY(data, in->data, in->size);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_sign(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    struct s2n_hash_state *digest = &sign->digest;
+
+    uint8_t digest_length = 0;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN] = { 0 };
+
+    RESULT_GUARD_POSIX(s2n_hash_digest_size(digest->alg, &digest_length));
+    RESULT_GUARD_POSIX(s2n_hash_digest(digest, digest_data, digest_length));
+
+    RESULT_ENSURE_LTE(digest_length, data_len);
+    
+    RESULT_CHECKED_MEMCPY(data, digest_data, digest_length);
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_async_pkey_op_set_output(struct s2n_async_pkey_op *op, const uint8_t *data, uint32_t data_len)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(data);
+
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    POSIX_GUARD_RESULT(s2n_async_get_actions(op->type, &actions));
+    POSIX_ENSURE_REF(actions);
+
+    POSIX_GUARD_RESULT(actions->set_output(op, data, data_len));
+    op->complete = true;
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_async_pkey_op_set_output_decrypt(struct s2n_async_pkey_op *op, const uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_decrypt_data *decrypt = &op->op.decrypt;
+    struct s2n_blob *out = &decrypt->decrypted;
+
+    RESULT_GUARD_POSIX(s2n_realloc(out, data_len));
+    RESULT_CHECKED_MEMCPY(out->data, data, data_len);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_op_set_output_sign(struct s2n_async_pkey_op *op, const uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    struct s2n_blob * sigcopy = &sign->signature;
```

Comment:
Thanks will audit the use of *. In this case, this function is a private API, and the OP pointer is checked in `s2n_async_pkey_op_set_output`.  Is there any policy for this? Should I always check the reference?

Reply:
Methods that start out private can become public later ;) It's safest to always check before de-referencing. This isn't production code though and it would probably just be annoying if CBMC started seg faulting, but we should still try to avoid that.

Reply:
You should always check the reference before dereferencing. Methods that start out private can become public later, and seg faults are not great.